src/analyses/variable-sensitivity/value_set_abstract_object.cpp

```diff
@@ -0,0 +1,305 @@
+/*******************************************************************\
+
+ Module: analyses variable-sensitivity
+
+ Author: diffblue
+
+\*******************************************************************/
+
+/// \file
+/// Value Set Abstract Object
+
+#include <analyses/variable-sensitivity/array_abstract_object.h>
+#include <analyses/variable-sensitivity/constant_abstract_value.h>
+#include <analyses/variable-sensitivity/constant_pointer_abstract_object.h>
+#include <analyses/variable-sensitivity/context_abstract_object.h>
+#include <analyses/variable-sensitivity/interval_abstract_value.h>
+#include <analyses/variable-sensitivity/struct_abstract_object.h>
+#include <analyses/variable-sensitivity/union_abstract_object.h>
+#include <analyses/variable-sensitivity/value_set_abstract_object.h>
+
+value_set_abstract_objectt::value_set_abstract_objectt(const typet &type)
+  : abstract_valuet(type), my_type(type_to_abstract_type(type))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(std::make_shared<constant_pointer_abstract_objectt>(type));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(std::make_shared<constant_abstract_valuet>(type));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+value_set_abstract_objectt::value_set_abstract_objectt(
+  const typet &type,
+  bool top,
+  bool bottom)
+  : abstract_valuet(type, top, bottom), my_type(type_to_abstract_type(type))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(
+      std::make_shared<constant_pointer_abstract_objectt>(type, top, bottom));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(
+      std::make_shared<constant_abstract_valuet>(type, top, bottom));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+value_set_abstract_objectt::value_set_abstract_objectt(
+  const exprt &expr,
+  const abstract_environmentt &environment,
+  const namespacet &ns)
+  : abstract_valuet(expr.type(), false, false),
+    my_type(type_to_abstract_type(expr.type()))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(std::make_shared<constant_pointer_abstract_objectt>(
+      expr, environment, ns));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(
+      std::make_shared<constant_abstract_valuet>(expr, environment, ns));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+abstract_object_pointert value_set_abstract_objectt::expression_transform(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  std::size_t num_operands = expr.operands().size();
+  PRECONDITION(operands.size() == num_operands);
+
+  std::vector<abstract_object_sett> collective_operands;
```

Comment:
These should just be `value_set_abstract_objectt`

Reply:
Because of the recursive nature of `apply_comb`, we need to be able to access the `n`th entry of `collective_operands`, and hence it can't just be a set.